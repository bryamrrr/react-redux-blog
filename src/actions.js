import api from './api';


function setPost(post) {
  return {
    type: 'SET_POST',
    payload: post,
  };
}


function setComments(comments) {
  return {
    type: 'SET_COMMENTS',
    payload: comments,
  };
}


function setUser(user) {
  return {
    type: 'SET_USER',
    payload: user,
  };
}


function postsNextPage() {
  return async (dispatch, getState) => {
    const state = getState();
    const currentPage = state.get('posts').get('page');

    const posts = await api.posts.getList(currentPage);

    dispatch(setPost(posts));

    return posts;
  };
}


function loadCommentsForPost(postId) {
  return async (dispatch) => {
    const comments = await api.posts.getComments(postId);
    dispatch(setComments(comments));
    return comments;
  };
}


function loadUser(userId) {
  return async (dispatch) => {
    const user = await api.users.getSingle(userId);
    dispatch(setUser(user));
    return user;
  };
}


export default {
  postsNextPage,
  loadCommentsForPost,
  loadUser,

  setPost,
  setComments,
  setUser,
};
