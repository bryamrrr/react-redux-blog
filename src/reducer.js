import { combineReducers } from 'redux-immutable';
import {
  fromJS,
  Map as map,
} from 'immutable';

const initialState = fromJS({
  posts: {
    page: 1,
    entities: {},
  },
  comments: {},
  users: {},
});


function postsPageReducer(state = initialState.get('posts').get('page'), action = {}) {
  switch (action.type) {
    case 'SET_POST': {
      return state + 1;
    }
    default: {
      return state;
    }
  }
}


function postsEntitiesReducer(state = initialState.get('posts').get('entities'), action = {}) {
  switch (action.type) {
    case 'SET_POST':
      return action.payload
        .reduce(
          (posts, post) => posts.set(post.id, map(post)),
          state,
        );
    default: {
      return state;
    }
  }
}


const postsReducer = combineReducers({
  page: postsPageReducer,
  entities: postsEntitiesReducer,
});


function commentsReducer(state = initialState.get('comments'), action = {}) {
  switch (action.type) {
    case 'SET_COMMENTS':
      return action.payload
        .reduce(
          (comments, comment) => comments.set(comment.id, map(comment)),
          state,
        );
    default: {
      return state;
    }
  }
}


function usersReducer(state = initialState.get('users'), action = {}) {
  switch (action.type) {
    case 'SET_USER':
      return state.set(action.payload.id, map(action.payload));
    default: {
      return state;
    }
  }
}


const reducer = combineReducers({
  posts: postsReducer,
  comments: commentsReducer,
  users: usersReducer,
});


export default reducer;
