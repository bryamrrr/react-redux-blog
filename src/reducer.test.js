import reducer from'./reducer';
import actions from'./actions';

test('reducer - SET_COMMENTS', () => {
  expect(
    reducer(
      undefined,
      actions.setComments([{ id: 1 }, { id: 3 }])
    )
  ).toMatchSnapshot();
});