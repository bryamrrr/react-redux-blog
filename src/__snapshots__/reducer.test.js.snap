// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer - SET_COMMENTS 1`] = `
Object {
  "comments": Object {
    "1": Object {
      "id": 1,
    },
    "3": Object {
      "id": 3,
    },
  },
  "posts": Object {
    "entities": Object {},
    "page": 1,
  },
  "users": Object {},
}
`;
